clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -9322,6 +9322,113 @@ static void handleOpenCLAccessAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) OpenCLAccessAttr(S.Context, AL));
 }
 
+static constexpr std::pair<Decl::Kind, StringRef>
+MakeDeclContextDesc(Decl::Kind K, StringRef SR) {
+  return std::pair<Decl::Kind, StringRef>{K, SR};
+}
+
+// FIXME: Can we perform a lookup instead?
+bool isDeviceAspectType(const QualType Ty) {
+  const EnumType *ET = Ty->getAs<EnumType>();
+  if (!ET)
+    return false;
+
+  std::array<std::pair<Decl::Kind, StringRef>, 3> Scopes = {
+      MakeDeclContextDesc(Decl::Kind::Namespace, "cl"),
+      MakeDeclContextDesc(Decl::Kind::Namespace, "sycl"),
+      MakeDeclContextDesc(Decl::Kind::Enum, "aspect")};
+
+  const auto *Ctx = cast<DeclContext>(ET->getDecl());
+  StringRef Name = "";
+
+  for (const auto &Scope : llvm::reverse(Scopes)) {
```

Comment:
I may have seen code this function contains somewhere around SemaSYCL. Should we avoid the duplication?

Reply:
Yep. I took the logic from there. I haven't looked too closely yet, but to avoid duplication I think we will need to move the `Util `class in `SemaSYCL `to `Sema `or something. Is it alright with you if I do that in a follow-patch?  I think it will add a lot of unrelated refactoring changes in `SemaSYCL `to this PR

Reply:
>  Is it alright with you if I do that in a follow-patch?
Sure!

Reply:
https://github.com/intel/llvm/issues/5186